File: dashboard/instrumentation/llm_metrics.py
"""
LLM token instrumentation — drop this file into any Python client project.

Usage
-----
1. pip install prometheus-client anthropic openai mistralai

2. Set the CLIENT_NAME env var to match the project slug used in Grafana
   (e.g. "metabelly", "client-b"). This is how tokens land under the right
   client in the dashboard.

3. Call track_anthropic / track_openai / track_mistral after each API call,
   or use the decorator versions.

4. Mount the /metrics endpoint (see bottom of file for FastAPI + Flask examples).
"""

import os
import functools
from prometheus_client import Counter, start_http_server, make_asgi_app

# ── The client label — set CLIENT_NAME in the app's environment ──────────────
CLIENT = os.environ.get("CLIENT_NAME", "unknown")

TOKENS = Counter(
    "api_tokens_used_total",
    "LLM API token consumption",
    ["client", "provider", "token_type", "model"],
)


def _inc(provider: str, model: str, input_tokens: int, output_tokens: int) -> None:
    TOKENS.labels(client=CLIENT, provider=provider, token_type="input",  model=model).inc(input_tokens)
    TOKENS.labels(client=CLIENT, provider=provider, token_type="output", model=model).inc(output_tokens)


# ── Anthropic ─────────────────────────────────────────────────────────────────

def track_anthropic(response) -> None:
    """Call after anthropic_client.messages.create(). Passes the response through."""
    _inc(
        provider="anthropic",
        model=response.model,
        input_tokens=response.usage.input_tokens,
        output_tokens=response.usage.output_tokens,
    )
    return response


def anthropic_create(client, **kwargs):
    """Drop-in wrapper for client.messages.create(**kwargs)."""
    response = client.messages.create(**kwargs)
    track_anthropic(response)
    return response


# ── OpenAI ────────────────────────────────────────────────────────────────────

def track_openai(response) -> None:
    """Call after openai_client.chat.completions.create(). Passes the response through."""
    _inc(
        provider="openai-mini" if "mini" in response.model else "openai",
        model=response.model,
        input_tokens=response.usage.prompt_tokens,
        output_tokens=response.usage.completion_tokens,
    )
    return response


def openai_create(client, **kwargs):
    """Drop-in wrapper for client.chat.completions.create(**kwargs)."""
    response = client.chat.completions.create(**kwargs)
    track_openai(response)
    return response


# ── Mistral ───────────────────────────────────────────────────────────────────

def track_mistral(response) -> None:
    """Call after mistral_client.chat.complete(). Passes the response through."""
    model = response.model or "mistral"
    provider = "mistral-small" if any(x in model for x in ("small", "nemo", "7b")) else "mistral"
    _inc(
        provider=provider,
        model=model,
        input_tokens=response.usage.prompt_tokens,
        output_tokens=response.usage.completion_tokens,
    )
    return response


def mistral_complete(client, **kwargs):
    """Drop-in wrapper for client.chat.complete(**kwargs)."""
    response = client.chat.complete(**kwargs)
    track_mistral(response)
    return response


# ── Mounting /metrics ─────────────────────────────────────────────────────────

def get_fastapi_metrics_app():
    """
    Mount on a FastAPI app:

        from llm_metrics import get_fastapi_metrics_app
        app.mount("/metrics", get_fastapi_metrics_app())
    """
    return make_asgi_app()


def start_standalone_metrics_server(port: int = 8000) -> None:
    """
    For non-ASGI apps (plain scripts, workers):

        from llm_metrics import start_standalone_metrics_server
        start_standalone_metrics_server(port=8000)

    Grafana Agent will scrape http://localhost:8000/metrics.
    """
    start_http_server(port)


# ── Flask example (commented out — uncomment if using Flask) ──────────────────
#
# from flask import Flask, Response
# from prometheus_client import generate_latest, CONTENT_TYPE_LATEST
#
# def register_flask_metrics(app: Flask) -> None:
#     @app.route("/metrics")
#     def metrics():
#         return Response(generate_latest(), mimetype=CONTENT_TYPE_LATEST)

File: dashboard/instrumentation/llmMetrics.ts
/**
 * LLM token instrumentation — drop this file into any Node.js/TypeScript client project.
 *
 * Setup
 * -----
 * npm install prom-client
 *
 * Set CLIENT_NAME env var to the project slug used in Grafana
 * (e.g. "metabelly", "client-b"). This is how tokens land under the
 * right client in the dashboard.
 *
 * Mount the /metrics route (examples at bottom of file).
 */

import { Counter, Registry, collectDefaultMetrics } from "prom-client";

const CLIENT = process.env.CLIENT_NAME ?? "unknown";

export const registry = new Registry();
collectDefaultMetrics({ register: registry });

const tokens = new Counter({
  name: "api_tokens_used_total",
  help: "LLM API token consumption",
  labelNames: ["client", "provider", "token_type", "model"] as const,
  registers: [registry],
});

function inc(provider: string, model: string, inputTokens: number, outputTokens: number) {
  tokens.labels({ client: CLIENT, provider, token_type: "input",  model }).inc(inputTokens);
  tokens.labels({ client: CLIENT, provider, token_type: "output", model }).inc(outputTokens);
}

// ── Anthropic ──────────────────────────────────────────────────────────────

/** Pass the raw response from anthropic.messages.create() */
export function trackAnthropic<T extends { model: string; usage: { input_tokens: number; output_tokens: number } }>(
  response: T
): T {
  inc("anthropic", response.model, response.usage.input_tokens, response.usage.output_tokens);
  return response;
}

// ── OpenAI ────────────────────────────────────────────────────────────────

/** Pass the raw response from openai.chat.completions.create() */
export function trackOpenAI<T extends { model: string; usage?: { prompt_tokens: number; completion_tokens: number } | null }>(
  response: T
): T {
  if (response.usage) {
    const provider = response.model.includes("mini") ? "openai-mini" : "openai";
    inc(provider, response.model, response.usage.prompt_tokens, response.usage.completion_tokens);
  }
  return response;
}

// ── Mistral ───────────────────────────────────────────────────────────────

/** Pass the raw response from mistral.chat.complete() */
export function trackMistral<T extends { model?: string | null; usage?: { promptTokens: number; completionTokens: number } | null }>(
  response: T
): T {
  if (response.usage) {
    const model = response.model ?? "mistral";
    const provider = /small|nemo|7b/i.test(model) ? "mistral-small" : "mistral";
    inc(provider, model, response.usage.promptTokens, response.usage.completionTokens);
  }
  return response;
}

// ── Mounting /metrics ──────────────────────────────────────────────────────

/**
 * Express / Next.js API route handler.
 *
 * Express:
 *   import { metricsHandler } from "./llmMetrics";
 *   app.get("/metrics", metricsHandler);
 *
 * Next.js app/api/metrics/route.ts:
 *   import { metricsHandler } from "@/lib/llmMetrics";
 *   export { metricsHandler as GET };
 */
export async function metricsHandler(
  _req: unknown,
  res: { setHeader: (k: string, v: string) => void; end: (body: string) => void }
) {
  res.setHeader("Content-Type", registry.contentType);
  res.end(await registry.metrics());
}

/**
 * Next.js App Router — use this as the GET export in app/api/metrics/route.ts:
 *
 *   import { nextMetricsRoute } from "@/lib/llmMetrics";
 *   export const GET = nextMetricsRoute;
 */
export async function nextMetricsRoute() {
  const { NextResponse } = await import("next/server");
  const body = await registry.metrics();
  return new NextResponse(body, {
    headers: { "Content-Type": registry.contentType },
  });
}
